### ISSUE TITLE
[Feature] Customizable condition for awaiting web server start

### ISSUE THREAD

**Issue by steinfurt** [microsoft/playwright#13535]
Hi,

thanks a lot for creating playwright - it is a joy to work with!

Regarding the webserver configuration, we have the following issue:

Our server has a status route "/status" which returns a 200 with payload `{ "state": "starting" }` or `{ "state": "started" }` depending on the exact state it is in. During the startup the app initiates several asynchronous tasks, e.g. it connects to a message broker which takes a while. So the app is already listening on the port, but not yet completely functional. For testing with playwright, we have to get from the "starting" phase to "started".

The current web server configuration does not support this case out-of-the-box, but it would be very convenient for us. The exact API would depend on how much freedom should be given. Two options which come to mind are

1. Response-based condition

```
webServer: {
  command: 'npm run start',
  url: 'http://localhost:3000/app/status',
  condition: (response) => response.status === 200 && response.data.state === 'started'
  timeout: 120 * 1000,
  reuseExistingServer: !process.env.CI,
}
```

thinking of an API-testing response object similar to what is returned from Axios GETs or `apiRequestContext.get`. So the type would be restricted to `(response: APIResponse) => boolean` and playwright would internally do a `const response = await axios.get(url)` and run `condition(response)` - of course without actually adding `axios` as a dependency.

2. Arbitrary condition

More generally, an arbitrary `condition: () => Promise<boolean>` in which users could perform more or less arbitrary checks. Users would then configure sth. like `condition: axios.get(url).then(response => response.status === 200 && response.data.state === 'started')`.

Obviously, there are workarounds: we could return status code 503 during the "starting" phase, start listening on the port only once the async tasks have finished or even start/wait for the web server before we run playwright, i.e. ignore the playwright webserver config option altogether. But having this generic option from playwright would be much more convenient.

Adding this functionality would, depending on the exact interface, also address #8206 and #12524. Also related: #13444

**Comment by rwoll** [comment]
Thanks—we'll keep this in mind when we work on Plugins. This will likely be addressed by #13362 directly by allowing you to give a custom poll function to the server plugin, or you'll be able to supply your own plugin.

**Comment by steinfurt** [comment]
Thanks, that sounds great! Looking forward to it...

**Comment by rmunn** [comment]
This, or the `readinessCheck: Promise<bool>` feature mentioned in https://github.com/microsoft/playwright/pull/13779#discussion_r861344756, would be a perfect fit for my situation, too: my webserver is launched in a Docker container via `docker-compose`, forwarding a port on localhost inside the container. Thing is, that creates a race condition, because Playwright is only watching the port on localhost. That port is opened as soon as docker-compose sets up the container, even though the webserver process inside the container may not be ready to listen on the port yet (as it's waiting for the DB container to be up before it starts responding on port 443).

Having some sort of configurable readiness check, or even just a URL that needs to return a 2xx status, would be enough to solve my use case, too.

**Comment by rwoll** [comment]
> Having some sort of configurable readiness check, or even just a URL that needs to return a 2xx status,

This is the behavior today (1.22): https://playwright.dev/docs/api/class-testconfig#test-config-web-server

1.23 will extend it to allow for more status codes.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you disagree with this closure, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.
